Fix for missing GC ref map record on ARM64 for the X8 register

trylek opened the pull request:
This is a simple hotfix for the generator only that works for me locally.
I suspect a similar adjustment needs to be done to the GC ref map parser.
We also need to bump up the ARM64 R2R PE format version as the images
emitted with this change will have an incompatible GC ref map compared
to the older ones.

Based on my offline chat with JanV I have decided to publish the WIP PR.
I'm looking forward to CoreCLR architects' suggestions on how to clean up
and finalize this change. One possibility for cleanup is to introduce a new
TransitionBlock function "OffsetOfFirstGCRefMapEntry" that will resolve to
OffsetOfArgumentRegisters on all architectures except for ARM64 where
it will be the offset of the X8 register (i.e. OffsetOfArgumentRegisters - 8).

Thanks

Tomas

trylek committed:
WIP: Fix for missing GC ref map record on ARM64 for the X8 register

This is a simple hotfix for the generator only that works for me locally.
I suspect a similar adjustment needs to be done to the GC ref map parser.
We also need to bump up the ARM64 R2R PE format version as the images
emitted with this change will have an incompatible GC ref map compared
to the older ones.

Thanks

Tomas

janvorli commented on src/vm/compile.cpp:
```diff
@@ -1007,13 +1007,19 @@ void CEECompileInfo::GetCallRefMap(CORINFO_METHOD_HANDLE hMethod, GCRefMapBuilde
     nStackSlots = nStackBytes / sizeof(TADDR) + NUM_ARGUMENT_REGISTERS;
 #else
-    nStackSlots = (sizeof(TransitionBlock) + nStackBytes - TransitionBlock::GetOffsetOfArgumentRegisters()) / TARGET_POINTER_SIZE;
+    nStackSlots = (sizeof(TransitionBlock) + nStackBytes - TransitionBlock::GetOffsetOfArgumentRegisters()) / TARGET_POINTER_SIZE + ReturnBuffPtrExtraSlot;
```
I would like to add a new TransitionBlock::OffsetOfFirstGCRefMapSlot function as you've suggested offline instead. It seems it would be easier to read the code that way.

briansull reviewed (commented):
comment

trylek committed:
Cleanup per Jan Vorlicek's suggestion

I have added a new TransitionBlock method GetOffsetOfFirstGCRefMapSlot
and I have fixed the consumption in PromoteCallerStackUsingGCRefMap.
I haven't yet addressed bumping up the R2R format version.

Thanks

Tomas

briansull reviewed (approved):
Looks pretty good now

trylek commented:
OK, I guess the last problem is deciding what to do about the R2R version. On a previous occasion @jkotas said a breaking format change should amount to incrementing READYTORUN_MAJOR_VERSION but I assume that means invalidating all pre-existing R2R images for all architectures. Is that appropriate?

AndyAyersMS reviewed (approved):
LGTM too -- have you had a chance to try the test case?

trylek commented:
I currently don't have any Linux ARM64 HW available for local testing, I just double-checked the GC ref map entry for the InitInvariant method.

AndyAyersMS commented:
Ok, let me pull your change and test it then.

trylek commented:
That would be just great, thank you so much!

janvorli reviewed (approved):
LGTM

AndyAyersMS commented:
Yep, looks like it fixes the test. Thanks!

trylek commented:
Great to hear, thank you! I plan to remove the WIP prefix from the PR title once I hear from Jan Kotas what needs to be done about the R2R format versioning.

xiangzhai commented:
:mips-interest
